### PR TITLE
bpo-41559: Implement PEP 612 - Add ParamSpec and Concatenate to typing

### PR DESCRIPTION
**Pull request by Fidget-Spinner**
Changes:

- [X] Add ParamSpec and Concatenate to typing.
- [X] Allow both in Callable
- [X] Allow ParamSpec in Generic 
- [X] Treat ParamSpec as a TypeVar  (add them to ``__parameters__`` ) in: 
  - [X] typing generics
  - [X] types.GenericAlias
- [X] Generics parameterized with ParamSpec and subscripted with lists should cast those lists to tuple:
  - [X] typing.Generic
  - [X] types.GenericAlias
- [X] Tests for ParamSpec
  - [X] Test specified use cases with typing.Callable.
  - [X] Test specified use cases with collections.abc.Callable.
  - [X] Test user defined generics
  - [X] TypeVar like substitution
- [X] Tests for Concatenate
- [X] Loosened Generic and GenericAlias type checks to allow for nearly anything.

Documentation will be another patch. 

https://bugs.python.org/issue41559

### CODE REVIEW

File: Lib/typing.py
Comment (gvanrossum) [inline Lib/typing.py]:
```suggestion
* _SpecialForm and its instances (special forms):
  Any, NoReturn, ClassVar, Union, Optional, Concatenate
* Classes whose instances can be type arguments in addition to types:
  ForwardRef, TypeVar, ParamSpec
```

File: Lib/typing.py
Comment (gvanrossum) [inline Lib/typing.py]:
```suggestion
        raise TypeError(f"Concatenate[...] is not valid as a type argument "
```
However, maybe it *is* valid? The PEP has examples like
```
def f(x: X[int, Concatenate[int, P_2]]) -> str: ...  # Accepted
```
I would also suggest not to be too strict -- we can leave that to the type checkers. At runtime we don't want to slow things down.

File: Lib/typing.py
Comment (gvanrossum) [inline Lib/typing.py]: Maybe add a reference to PEP 612 and an example, e.g. `Callable[Concatenate[int, P], int]`? (No need to explain `P`.)

File: Lib/typing.py
Comment (gvanrossum) [inline Lib/typing.py]: According to PEP 612, the final parameter must be a `ParamSpec` (I can't tell whether other parameters may also be `ParamSpec`.)

File: Lib/typing.py
Comment (gvanrossum) [inline Lib/typing.py]: Could this perhaps be `__init__()`?

File: Lib/typing.py
Comment (gvanrossum) [inline Lib/typing.py]:
```suggestion
        self._setup_bound_cov_contra(bound, covariant, contravariant)
```

File: Lib/typing.py
Comment (gvanrossum) [inline Lib/typing.py]:
```suggestion
    Parameter specification variables exist primarily for the benefit of static
    type checkers.  They are used to forward the parameter types of one
```

File: Lib/typing.py
Comment (gvanrossum) [inline Lib/typing.py]:
```suggestion
           '''A type-safe decorator to add logging to a function.'''
```

File: Lib/test/test_typing.py
Comment (gvanrossum) [inline Lib/test/test_typing.py]: I'm confused. If C is generic over T and P, how is C[int] even valid? In general I'd rather not make this test dependent on ParamSpec -- I'd rather add a new test for that. Same below.

**Review by gvanrossum — COMMENTED** [review]
I'm sorry, I didn't get to the end of this review (not by far), but I have a few comments. Main thing though is, did you look at all the examples in PEP 612? It seems that for user-defined generic classes that have a ParamSpec parameter, things like `X[int, P]` are actually valid.

Maybe this indirectly also answers your question: I think that the ParamSpec should be included in `__parameters__`.

Also, I wonder if we could do less type checking at runtime rather than trying to be strict? I desperately want typing.py to shrink, not grow... :-)

**Comment by Fidget-Spinner** [comment]
Points taken. Sorry ``X[int, P]`` slipped my mind since I hadn't fixed ``__parameters__`` yet. 

D'oh, I just realised I answered my own question since we need ``__parameters__`` to implement ``X[int, P]`` in the first place.

**Comment by gvanrossum** [comment]
Let's wait until #23060 has landed and then see what adjustments this might need.

**Comment by Fidget-Spinner** [comment]
This should now be ready now @gvanrossum. The implementation was longer than I wanted it to be, but I wanted to at least maintain correct ``__args__`` during runtime.  It should work for all cases the PEP defined, but it falls apart quite wildly when used in unintended ways due to the removal of most runtime checking. Eg:

``typing.Callable[P, P][[int, str], [int, str]]`` yields the clearly wrong ``typing.Callable[[[<class 'int'>, <class 'str'>], [<class 'int'>, <class 'str'>], [<class 'int'>, <class 'str'>]], [<class 'int'>, <class 'str'>]]``. While Callable[P, T][[int, str], [int, str]] yields the correct ``typing.Callable[[int, str], [<class 'int'>, <class 'str'>]]``. 

Both expressions are currently invalid, but I wanted to point that out since it seemed interesting to me anyways :).

PEP 612's grammar states:
```
callable ::= Callable "[" parameters_expression, type_expression "]"

parameters_expression ::=
  | "..."
  | "[" [ type_expression ("," type_expression)* ] "]"
  | parameter_specification_variable
  | concatenate "["
                   type_expression ("," type_expression)* ","
                   parameter_specification_variable
                "]"
```
> As before, parameters_expressions by themselves are not acceptable in places where a type is expected

**Comment by Fidget-Spinner** [comment]
Actually I just realized PEP 612 doesn't explicitly say that ParamSpec should support TypeVar-like substitution. Maybe I over complicated things a little by allowing that in ``types.GenericAlias``?

**Comment by gvanrossum** [comment]
Finally looking at this...

> Actually I just realized PEP 612 doesn't explicitly say that ParamSpec should support TypeVar-like substitution. Maybe I over complicated things a little by allowing that in `types.GenericAlias` (just for collections.abc.Callable)?

Yeah, ParamSpec is super special, and you can't do anything with it that isn't explicitly mentioned in the PEP. So it's debatable whether `Callable[P, int]` should even have `P` in its `__parameters__`. Then again, for the simple case `A = Callable[P, int]` there's no reason why we couldn't allow `A[[str, str]]` and get `Callable[[str, str], int]` as the result. But type checkers wouldn't know what to do with that -- ParamSpec is a magic cookie.

So let's not worry about those clearly wrong cases.

File: Lib/_collections_abc.py
Comment (gvanrossum) [inline Lib/_collections_abc.py]: How about we refactor things a bit more so the helper function does all three checks? The helper could be called `_has_special_args()` and it would return detect all three cases: `[..., rt]`, `[P, rt]` and `[Concatenate[...], rt]`. You could then also get rid of `_is_typing_names()`, which is called only once.

File: Lib/_collections_abc.py
Comment (gvanrossum) [inline Lib/_collections_abc.py:465]: Add a comment explaining that args[0] occurs due to things like Z[[int, str, bool]] (this took me a while to figure out :-).

File: Lib/test/test_typing.py
Comment (gvanrossum) [inline Lib/test/test_typing.py]: These seem to have the same name?

File: Lib/test/test_typing.py
Comment (gvanrossum) [inline Lib/test/test_typing.py]: Well, ParamSpec is not allowed to occur in Union.

**Comment by gvanrossum** [comment]
Sorry, submitted too soon.

File: Lib/test/test_typing.py
Comment (gvanrossum) [inline Lib/test/test_typing.py]: These error cases seem fairly unimportant.

File: Lib/test/test_typing.py
Comment (gvanrossum) [inline Lib/test/test_typing.py]: This comment is confusing -- maybe "Used to raise TypeError"?

File: Lib/typing.py
Comment (gvanrossum) [inline Lib/typing.py]: I wonder if things would become easier if we just treated ParamSpec as a subclass of TypeVar? Then it's just a type variable except in those few cases where it actually needs to behave differently at runtime (maybe only the weird exception in the PEP for `Z[[int, int]] == Z[int, int]`?). Might save on tests as well.

File: Lib/typing.py
Comment (gvanrossum) [inline Lib/typing.py]:
```suggestion
        return (params,)
```

File: Lib/typing.py
Comment (gvanrossum) [inline Lib/typing.py:562]: I'd check this before the loop calling `_type_check(p, msg)`.

File: Lib/_collections_abc.py
Comment (gvanrossum) [inline Lib/_collections_abc.py]: PS. Above on line 419 there's a space missing between ``` flattened``argtypes`` ```.

File: Objects/genericaliasobject.c
Comment (gvanrossum) [inline Objects/genericaliasobject.c]: The `&=` operator worries me, since the return value of `strcmp()` can be `-1`, `0` or `1`. I don't want to have to reason about the value of `(-1) & 1`. And why not just write something like
```
if (strcmp(...) != 0)
    return 0;
```
? Then you don't even need the `cmp` variable.

File: Objects/genericaliasobject.c
Comment (gvanrossum) [inline Objects/genericaliasobject.c]: Why only do this for ParamSpec?

File: Objects/genericaliasobject.c
Comment (gvanrossum) [inline Objects/genericaliasobject.c:352]: (OT: I hate that special case. There's a Zen of Python line about special cases. :-)

**Review by gvanrossum — COMMENTED** [review]
Excellent work! Thanks for your perseverance. My feedback is mostly simple improvements.

File: Objects/genericaliasobject.c
Comment (Fidget-Spinner) [reply Objects/genericaliasobject.c]: I'm afraid of existing code relying on substituting typevars in genericalias with lists, then expecting that list to be inside `__args__`. Although in Python 3.9 that should be invalid, and the IDE/type checker should warn the users about that already.

If I just convert list to tuples for all cases, the C code will become way simpler :).

File: Objects/genericaliasobject.c
Comment (Fidget-Spinner) [reply Objects/genericaliasobject.c]: That would make checking for paramspec fail because the equivalent python code is (please pardon the minimal example):
```python
def is_typevar_like(obj):
    names = ('TypeVar', 'ParamSpec')
    for name in names:
        if x!= name:
            return False

# False
is_typevar_like('ParamSpec')
```
I used `&=` since 0 short-circuits (unsure if that's the correct term) the entire & operation to 0 as long as there's a single 0 (single match) in there. ``(-1) & 1`` will always give ``1``. That's the closest equivalent to python's `any` I could think of.

File: Objects/genericaliasobject.c
Comment (Fidget-Spinner) [reply Objects/genericaliasobject.c:352]: Same here 😉 , sadly this is one special case that seems special enough to break the rules :(.

File: Objects/genericaliasobject.c
Comment (gvanrossum) [reply Objects/genericaliasobject.c]: There can't be much code yet relying about anything for GenericAlias -- it's new in 3.9. We could backport that particular behavior to 3.9.3 to ensure there won't be any more code relying on it. :-)

File: Objects/genericaliasobject.c
Comment (gvanrossum) [reply Objects/genericaliasobject.c]: Oh, I didn't see that your 'for' condition skips if `cmp` is zero. Sorry!

But that's kind of easy to overlook. How about this way:
```
    for (int i = 0; i < num; ++i) {
        if (strcmp(type->tp_name, va_arg(names, const char *)) == 0) {
            cmp = 0;
            break;
        }
    }
```
Then I would rename `cmp` to `eq` so the reader understands it represents "equality" or perhaps to `hit` meaning "we found the name in the list of names" ("hit" is somewhat traditional for a search loop, which this is).

Finally I would change the function name to `is_typing_name()` (i.e., singular). This again helps the reader get the purpose of the function.

File: Lib/typing.py
Comment (gvanrossum) [reply Lib/typing.py]: Thoughts about this?

File: Lib/_collections_abc.py
Comment (Fidget-Spinner) [reply Lib/_collections_abc.py:465]: Oh oops I just remembered one more reason why I had to do this: 
After we subclassed `types.GenericAlias` for Callable in the other PR, I forgot to override the ``__getitem__``, so it automatically returned a plain `types.GenericAlias` rather than the special subclass during typevar substitution. The problem with that is that the `repr` is now wrong, as it shows the underlying flattened version rather than the one the user writes.

Eg.
```python
>>> Callable[[dict, float], T][int]
collections.abc.Callable[dict, float, int]
```
We probably need to backport some part of this change to the 3.9 branch to fix that.

File: Lib/typing.py
Comment (Fidget-Spinner) [reply Lib/typing.py]: I changed all the methods that check for `(TypeVar, ParamSpec)` to just check for `_TypeVarLike` instead. Which means I can probably get rid of some tests for `ParamSpec` since it uses the exact same `__reduce__`, `__repr__`, `__ror__` etc. methods as `TypeVar`, and `TypeVar` already has tests for them.

File: Lib/_collections_abc.py
Comment (Fidget-Spinner) [reply Lib/_collections_abc.py:467]: Ouch, i just noticed something else, the repr for both Callables in `typing` and `collections.abc` don't show `types.GenericAlias` properly either:
```python
>>> typing.Callable[dict[float, str], int]
typing.Callable[[dict[float, str]], int]
>>> Callable[dict[float, str], int]
collections.abc.Callable[[dict[float, str]], int]
```
Looks like something else we need to backport to 3.9

File: Lib/_collections_abc.py
Comment (gvanrossum) [inline Lib/_collections_abc.py]: If you want to you can move the `len() == 2` test into the helper too (just pass args instead of args[0]).

**Review by gvanrossum — COMMENTED** [review]
Re: the 3.9 issues that you discovered: maybe create a PR for those first, land that in master and 3.9 branch, then merge master into this PR?

**Comment by Fidget-Spinner** [comment]
Guido, thank you so much for all the time you've put into reviewing this PR! I learnt quite a bit about the typing module in the process :).

> Re: the 3.9 issues that you discovered: maybe create a PR for those first, land that in master and 3.9 branch, then merge master into this PR?
Yeah I agree, I created #23915 for that. Once that one lands, I'll merge the changes in along with some of your suggestions.

File: Lib/_collections_abc.py
Comment (gvanrossum) [inline Lib/_collections_abc.py]: More idiomatic:
```suggestion
    if len(args) != 2:
```

**Review by gvanrossum — APPROVED** [review]

**Comment by gvanrossum** [comment]
I think it’s ready!

**Comment by gvanrossum** [comment]
@Fidget-Spinner Let me know if you agree.

**Comment by Fidget-Spinner** [comment]
> @Fidget-Spinner Let me know if you agree.

Yes! Sorry, I thought you were going to merge it in so I was waiting too 😆 .

**Comment by gvanrossum** [comment]
I figured that was the case. :-)

**Comment by Fidget-Spinner** [comment]
@gvanrossum  Oh dang I knew I had a hunch I missed something :/. I forgot to convert all lists to tuples in ga_new/setup_ga too (I did it only for getitem and co.). Well on the bright side, having that as a separate PR makes backporting to 3.9 easier if anything :), and anyways lists aren't valid in genericalias other than the Callable case (which we already have covered), so we won't have anything broken even without that in *right now*.

**Comment by Fidget-Spinner** [comment]
I'll fix that some other time, right now I'm celebrating 🥳

**Pull request merged**
